sigstore/_internal/trustroot.py

```diff
@@ -58,48 +76,192 @@ def _is_timerange_valid(period: TimeRange | None, *, allow_expired: bool) -> boo
+
+    This structure exists to facilitate key rotation in a CT log.
+    """
+
+    def __init__(self, keys: List[bytes] = []):
+        """
+        Create a new `Keyring`, with `keys` as the initial set of signing
+        keys. These `keys` can be in either DER format or PEM encoded.
+        """
+        self._keyring = {}
+        for key_bytes in keys:
+            key = None
+
+            try:
+                key = load_pem_public_key(key_bytes)
+            except UnexpectedKeyFormatError as e:
+                raise e
+            except InvalidKeyError:
+                key = load_der_public_key(key_bytes)
+
+            self._keyring[key_id(key)] = key
+
+    def add(self, key_pem: bytes) -> None:
+        """
+        Adds a PEM-encoded key to the current keyring.
+        """
+        key = load_pem_public_key(key_pem)
+        self._keyring[key_id(key)] = key
+
+    def verify(self, *, key_id: KeyID, signature: bytes, data: bytes) -> None:
+        """
+        Verify that `signature` is a valid signature for `data`, using the
+        key identified by `key_id`.
+
+        Raises if `key_id` does not match a key in the `Keyring`, or if
+        the signature is invalid.
+        """
+        key = self._keyring.get(key_id)
+        if key is None:
+            # If we don't have a key corresponding to this key ID, we can't
+            # possibly verify the signature.
+            raise KeyringLookupError(f"no known key for key ID {key_id.hex()}")
+
+        try:
+            if isinstance(key, rsa.RSAPublicKey):
+                key.verify(
+                    signature=signature,
+                    data=data,
+                    padding=padding.PKCS1v15(),
+                    algorithm=hashes.SHA256(),
+                )
+            elif isinstance(key, ec.EllipticCurvePublicKey):
+                key.verify(
+                    signature=signature,
+                    data=data,
+                    signature_algorithm=ec.ECDSA(hashes.SHA256()),
+                )
+            else:
+                # NOTE(ww): Unreachable without API misuse.
+                raise KeyringError(f"unsupported key type: {key}")
+        except InvalidSignature as exc:
+            raise KeyringSignatureError("invalid signature") from exc
+
+
+RekorKeyring = NewType("RekorKeyring", Keyring)
+CTKeyring = NewType("CTKeyring", Keyring)
+
+
+class KeyringPurpose(str, Enum):
+    """
+    Keyring purpose typing
+    """
+
+    SIGN = "sign"
+    VERIFY = "verify"
+
+    def __str__(self) -> str:
+        """Returns the purpose string value."""
+        return self.value
+
+
 class TrustedRoot(_TrustedRoot):
     """Complete set of trusted entities for a Sigstore client"""
 
+    args: Optional[Namespace] = None
+    purpose: KeyringPurpose
+
     @classmethod
-    def from_file(cls, path: str) -> "TrustedRoot":
+    def from_file(
+        cls,
+        path: str,
+        args: Optional[Namespace] = None,
```

Comment:
This looks off -- I don't think we want to push the CLI state into the `TrustedRoot` like this. I _think_ we want to remove this entirely, right?

Reply:
Yes, you're right! I did that as a workaround for having args, but it seems weird and all args should be removed.